### PR TITLE
8341781: Improve Min/Max node identities

### PR DESCRIPTION
**Pull request by jaskarth**
Hi all,
This patch implements some missing identities for Min/Max nodes. It adds static type-based operand choosing for MinI/MaxI, such as the ones that MinL/MaxL use. In addition, it adds simplification for patterns such as `Max(A, Max(A, B))` to `Max(A, B)` and `Max(A, Min(A, B))` to `A`. These simplifications stem from the [lattice identity rules](https://en.wikipedia.org/wiki/Lattice_(order)#As_algebraic_structure). The main place I've seen this pattern is with MinL/MaxL nodes created during loop optimizations. Some examples of where this occurs include BigInteger addition/subtraction, and regex code. I've run some of the existing benchmarks and found some nice improvements:
```c
                                                                Baseline                    Patch
Benchmark                                 Mode  Cnt       Score       Error  Units    Score       Error  Units  Improvement
BigIntegers.testAdd                       avgt   15      25.096 ±     3.936  ns/op   19.214  ±    0.521  ns/op  (+ 26.5%)
PatternBench.charPatternCompile           avgt    8     453.727 ±   117.265  ns/op   370.054 ±   26.106  ns/op  (+ 20.3%)
PatternBench.charPatternMatch             avgt    8     917.604 ±   121.766  ns/op   810.560 ±   38.437  ns/op  (+ 12.3%)
PatternBench.charPatternMatchWithCompile  avgt    8    1477.703 ±   255.783  ns/op  1224.460 ±   28.220  ns/op  (+ 18.7%)
PatternBench.longStringGraphemeMatches    avgt    8     860.909 ±   124.661  ns/op   743.729 ±   22.877  ns/op  (+ 14.6%)
PatternBench.splitFlags                   avgt    8     420.506 ±    76.252  ns/op   321.911 ±   11.661  ns/op  (+ 26.6%)
```
I've added some IR tests, and tier 1 testing passes on my linux machine. Reviews would be appreciated!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341781](https://bugs.openjdk.org/browse/JDK-8341781): Improve Min/Max node identities (**Enhancement** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21439/head:pull/21439` \
`$ git checkout pull/21439`

Update a local copy of the PR: \
`$ git checkout pull/21439` \
`$ git pull https://git.openjdk.org/jdk.git pull/21439/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21439`

View PR using the GUI difftool: \
`$ git pr show -t 21439`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21439.diff">https://git.openjdk.org/jdk/pull/21439.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21439#issuecomment-2403845254)
</details>

### CODE REVIEW

File: src/hotspot/share/opto/addnode.hpp
Comment (liach) [inline src/hotspot/share/opto/addnode.hpp:270]: The old comment above 
```
// all the behavior of addition on a ring.  Only new thing is that we allow
// 2 equal inputs to be equal.
```
seems outdated.

File: test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:122]: Since `MinL/MaxL` are expanded in macro expansion, this rule will also succeed even if the optimization is not applied. I suggest to also add `phase = CompilePhase.BEFORE_MACRO_EXPANSION`. Same below.

File: test/hotspot/jtreg/compiler/vectorization/runner/BasicShortOpTest.java
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/vectorization/runner/BasicShortOpTest.java:220]: I guess without this change, this collapses to a constant which enables vectorization which was not expected before?

**Review by chhagedorn — COMMENTED** [review]
Few comments, otherwise, looks good to me.

**Comment by chhagedorn** [comment]
Your new test fails on Linux with `-XX:UseAVX=0`:

```
One or more @IR rules failed:

Failed IR Rules (8) of Methods (8)
----------------------------------
1) Method "public double compiler.c2.irTests.TestMinMaxIdentities.doubleMaxMax(double,double)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MAX_D#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MaxD.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

2) Method "public double compiler.c2.irTests.TestMinMaxIdentities.doubleMaxMin(double,double)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MIN_D#_", "1", "_#MAX_D#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MinD.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
         * Constraint 2: "(\\d+(\\s){2}(MaxD.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

3) Method "public double compiler.c2.irTests.TestMinMaxIdentities.doubleMinMax(double,double)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MIN_D#_", "1", "_#MAX_D#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MinD.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
         * Constraint 2: "(\\d+(\\s){2}(MaxD.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

4) Method "public double compiler.c2.irTests.TestMinMaxIdentities.doubleMinMin(double,double)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MIN_D#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MinD.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

5) Method "public float compiler.c2.irTests.TestMinMaxIdentities.floatMaxMax(float,float)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MAX_F#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MaxF.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

6) Method "public float compiler.c2.irTests.TestMinMaxIdentities.floatMaxMin(float,float)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MIN_F#_", "1", "_#MAX_F#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MinF.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
         * Constraint 2: "(\\d+(\\s){2}(MaxF.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

7) Method "public float compiler.c2.irTests.TestMinMaxIdentities.floatMinMax(float,float)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MIN_F#_", "1", "_#MAX_F#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MinF.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
         * Constraint 2: "(\\d+(\\s){2}(MaxF.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

8) Method "public float compiler.c2.irTests.TestMinMaxIdentities.floatMinMin(float,float)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MIN_F#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(MinF.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
```
Looks like we do not emit `Min/MaxF/D` nodes with `UseAVX=0`. I quickly checked the code and indeed, the intrinsics are only enabled if `UseAVX >= 1`:
https://github.com/openjdk/jdk/blob/16042556f394adfa93e54173944198397ad29dea/src/hotspot/cpu/x86/x86.ad#L1542-L1549

You can probably just update your tests to exclude IR matching for this setup. Maybe you also want to double check the other architectures.

File: test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:116]: This is a good call, the IR rules don't apply to RISC-V because it doesn't have support for CMoves so the MinL/MaxL nodes aren't made at all. Since `Math.min/max(LL)` isn't intensified it first needs to be matched into CMove, then Min/Max, and then the identity needs to be called. Since #20098 implements the intrinsic we could remove the special casing after it's merged. I've added a comment to the source code as well.

File: test/hotspot/jtreg/compiler/vectorization/runner/BasicShortOpTest.java
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/vectorization/runner/BasicShortOpTest.java:220]: Yeah, exactly - since `65536` is larger than the maximum short value, with this patch it can optimize the MinI node away entirely. I changed it to use the `b` array, which is what the `vectorMax` test case below uses.

File: test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:116]: On closer inspection it seems that because of the CMove cost model, the outer min/max operation doesn't get turned into a CMove so the Long IR rules don't reliably get matched anywhere. It must have slipped through the cracks because of the way that my IR rules were structured, I only realized this after I added the compile phase to the other 2 rules. I think for this to work it would need the intrinsics from the other PR. Do you think we should continue with this PR with the Long cases disabled and enable it afterwards, or should we wait for #20098 to be merged?

Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:116]: Thanks for sharing more details. I think it's perfectly fine to still add them now but leave them disabled with a reference to JDK-8307513 since you already wrote them.

Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:116]: Sounds good!

**Comment by jaskarth** [comment]
Thanks for the suggestions and testing, @liach and @chhagedorn! I've taken a look at the backend implementations, and it seems that aarch64 and RISC-V unconditionally support floating point Min/Max while x64 only supports them with `UseAVX >= 1`, as described. I made it so that the test only runs when it matches that criteria. I've pushed a commit that should address all the suggestions here.

**Review by chhagedorn — APPROVED** [review]
Looks good, thanks for the update! I'll give this another spinning in our testing.

**Comment by chhagedorn** [comment]
Testing passed!

File: src/hotspot/share/opto/addnode.cpp
Comment (jaskarth) [reply src/hotspot/share/opto/addnode.cpp:1268]: I think we can't encounter `TOP` here because we run `Value()` before `Identity()`, so if TOP is returned by Value() the idealization process exits to return the top node before running Identity().

File: test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: From my understanding this isn't possible as-is since CPU features seem to be checked regardless of whether the architecture supports it or not, so we can't simply check for AVX because that would fail on aarch64 and riscv64. I think we could work around this with `applyIfCPUFeatureOr = {"avx", "true", "asimd", "true", "rvv", "true"}` to force a check for all 3 platforms but it'd be filtering more platforms than strictly necessary.

Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: Which platforms would be filtered "more platforms than strictly necessary"?

Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: With the workaround to check for CPU features on all 3 platforms, we'd be not checking the IR when `asimd = false` or `rvv = false`, but the IR check should pass with those features too.

Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: I think your tags are actually available, with `applyIfPlatform`. Look at `irTestingPlatforms` in https://github.com/openjdk/jdk/blame/master/test/hotspot/jtreg/compiler/lib/ir_framework/test/IREncodingPrinter.java#L69

Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: Example:
`./test/hotspot/jtreg/compiler/loopopts/superword/TestGeneralizedReductions.java:    @IR(applyIfPlatform = {"riscv64", "true"},`

Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: The usage of these is quite rare - usually we focus more on the CPU features, and not the platform tags ;)

Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: Ah, I was meaning that with `applyIfPlatformOr = {"x64", "true", "aarch64", "true", "riscv64", "true"}` we would still need at least `applyIfCPUFeature = {"avx", true"}` because on x86 we only make Min/MaxF and Min/MaxD with AVX. But that applyIfCPUFeature will check for AVX on aarch64 and riscv64 as well, but will fail because it's not available for those platforms. That's why I suggested the workaround of checking the other CPU features, to make the test at least run on the other platforms. It'd be nice to be able to express platform and CPU feature combinations like with `@requires`, but the use-case here is pretty niche.

Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: First: yes, I'm fine with just using CPU features - it will IR test it on fewer platforms than maybe desired, but that is ok, I think.

I guess the issue with `@IR` is that it only allows `AND` or `OR` clauses ... and not the more complicated mix of `&` and `|` from `@requires`. But in theory, you can just have multiple `@IR` rules (that simulates the OR):
- One for `x64` (applyIfPlatform) and `avx` (applyIfCPUFeature)
- One IR rule for each of: `aarch64` and `riscv64` (applyIfPlatform)

But again, I'm ok with only checking for CPU features... it is more simple.

Comment (eme64) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:35]: @chhagedorn you may be interested in this conversation as well ;)

Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java]: I hadn't considered using multiple IR checks, that would definitely work as well. I've pushed a commit that just uses the CPU features, to keep it simple. Let me know what you think!

File: test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java
Comment (eme64) [inline test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java:120]: I would say you should make them negative for now, i.e. make them `failOn`. Otherwise we won't catch these cases when JDK-8307513 gets integrated ;)

**Comment by eme64** [comment]
The IR rules look ok to me. Nice progress :)

File: test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java]: Sounds good, I've pushed a commit that makes the tests pass now but fail when 8307513 is integrated.

**Review by chhagedorn — APPROVED** [review]
The IR rule updates look good. Maybe you want to wait with integrating this until after the fork next Thursday. So, this only goes into JDK 25.

**Comment by jaskarth** [comment]
Thanks for the re-review! I think that's a good idea, I'll integrate it after JDK 24 is forked.

**Comment by jaskarth** [comment]
Since it's after the fork, I'll integrate it now. Thanks again for the reviews everyone!
/integrate

**Pull request closed without merging**
